adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -329,39 +390,39 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
                     final Future<Device> result = Future.future();
                     usernamePasswordAuthProvider.authenticate(credentials, result.completer());
                     return result;
-                }).map(authenticatedDevice -> {
+                }).compose(authenticatedDevice -> {
                     LOG.debug("successfully authenticated device [tenant-id: {}, auth-id: {}, device-id: {}]",
                             authenticatedDevice.getTenantId(), credentials.getAuthId(),
                             authenticatedDevice.getDeviceId());
                     onAuthenticationSuccess(endpoint, authenticatedDevice);
-                    return null;
-                }).otherwise(t -> {
+                    return accepted ( authenticatedDevice );
+                }).recover(t -> {
                     LOG.debug("cannot authenticate device [tenant-id: {}, auth-id: {}]",
                             credentials.getTenantId(), credentials.getAuthId(), t);
-                    if (ServerErrorException.class.isInstance(t)) {
+                    if (t instanceof ServerErrorException) {
                         // one of the services we depend on might not be available (yet)
-                        endpoint.reject(MqttConnectReturnCode.CONNECTION_REFUSED_SERVER_UNAVAILABLE);
+                        return rejected(CONNECTION_REFUSED_SERVER_UNAVAILABLE);
                     } else {
                         // validation of credentials has failed
-                        endpoint.reject(MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
+                        return rejected(CONNECTION_REFUSED_NOT_AUTHORIZED);
                     }
-                    return null;
                 });
+
             }
         }
     }
 
     private void onAuthenticationSuccess(final MqttEndpoint endpoint, final Device authenticatedDevice) {
 
         endpoint.closeHandler(v -> {
-            close(endpoint);
+            close(endpoint, authenticatedDevice);
             LOG.debug("connection to device [tenant-id: {}, device-id: {}] closed",
                     authenticatedDevice.getTenantId(), authenticatedDevice.getDeviceId());
             metrics.decrementMqttConnections(authenticatedDevice.getTenantId());
+            sendDisconnectedEvent(endpoint.clientIdentifier(), authenticatedDevice);
```

Comment:
looks like you are sending the disconnected twice, once as part of *close()* and once explicitly.

Reply:
Indeed. Fixed!